db/sg_replicate_util.go

```diff
@@ -12,36 +12,70 @@ package db
+//     {"channels": ["channel1", "channel2"] }
+//
+// 3. As a JSON object with a property for each collection, with the value being an array of channels:
+//
+//	{
+//	  "collections_channels": {
+//	    "collection1": ["scope1.channel1"],
+//	    "collection2": ["scope1.channel1", "scope1.channel2"]
+//	  }
+//	}
+func CollectionChannelsFromQueryParams(namedCollections []string, queryParams interface{}) (perCollectionChannels [][]string, allCollectionsChannels []string, err error) {
+	switch val := queryParams.(type) {
+	case map[string]interface{}:
+		if chanarray, ok := val["channels"].([]interface{}); ok {
+			allCollectionsChannels, err = interfaceValsToStringVals(chanarray)
+			return nil, allCollectionsChannels, err
+		}
 
-	var chanarray []interface{}
-	if paramsmap, ok := queryParams.(map[string]interface{}); ok {
-		if chanarray, ok = paramsmap["channels"].([]interface{}); !ok {
-			return nil, errors.New("Replication specifies sync_gateway/bychannel filter, but query_params is missing channels property")
+		if collectionChannels, ok := val["collections_channels"].(map[string]interface{}); ok {
+			perCollectionChannels = make([][]string, len(namedCollections))
+			for i, collection := range namedCollections {
+				collectionQueryParams := collectionChannels[collection]
+				collectionQueryParamsArray, ok := collectionQueryParams.([]interface{})
+				if !ok {
+					return nil, nil, errors.New("query_params must be a map of collection name to array of channels")
+				}
+				channels, err := interfaceValsToStringVals(collectionQueryParamsArray)
+				if err != nil {
+					return nil, nil, err
+				}
+				perCollectionChannels[i] = channels
+			}
 		}
-	} else if chanarray, ok = queryParams.([]interface{}); ok {
-		// query params is an array and chanarray has been set, now drop out of if-then-else for processing
-	} else {
-		return nil, base.HTTPErrorf(http.StatusBadRequest, ConfigErrorBadChannelsArray)
+		return perCollectionChannels, nil, nil
+	case []interface{}:
+		allCollectionsChannels, err = interfaceValsToStringVals(val)
+		return nil, allCollectionsChannels, err
+	default:
+		return nil, nil, base.HTTPErrorf(http.StatusBadRequest, ConfigErrorBadChannelsArray)
 	}
-	if len(chanarray) > 0 {
-		channels = make([]string, len(chanarray))
-		for i := range chanarray {
-			if channel, ok := chanarray[i].(string); ok {
-				channels[i] = channel
-			} else {
-				return nil, errors.New("Bad channel name in query_params for sync_gateway/bychannel filter")
-			}
+}
+
+// interfaceValsToStringVals takes a []interface{} and returns a []string
+func interfaceValsToStringVals(interfaceVals []interface{}) ([]string, error) {
+	stringVals := make([]string, len(interfaceVals))
+	for i := range interfaceVals {
+		str, ok := interfaceVals[i].(string)
+		if !ok {
+			return nil, fmt.Errorf("Bad channel name %v (%T) in query_params for sync_gateway/bychannel filter - expecting strings", interfaceVals[i], interfaceVals[i])
```

Comment:
do we need `base.UD` here, or is it OK because it will probably only make it back to user again?